File: leetcode/39.combination-sum.cpp
class Solution {
private:
    void helper(vector<int>& candidates, int start, int sum, int target, vector<int>& ans, vector<vector<int>>& res)
    {
        if (sum >= target)
        {
            if (sum == target)
            {
                res.push_back(ans);
            }
            return;
        }
        
        for (auto i = start; i < candidates.size(); ++i)
        {
            ans.push_back(candidates[i]);
            helper(candidates, i, sum + candidates[i], target, ans, res);
            ans.pop_back();
        }
        
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> ans;
        
        helper(candidates, 0, 0, target, ans, res);
        return res;
    }
};

File: leetcode/39.combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        ans = []
        candidates.sort()
        self.helper(candidates, 0, target, ans, 0, res)
        
        return res
        
    def helper(self, candidates, start, target, ans, total, res):
        if total == target:
            res.append(list(ans))
            return
            
        for i in range(start, len(candidates)):
            
            number = candidates[i]
            
            if total + number <= target:
                ans.append(number)
                total += number
                
                self.helper(candidates, i, target, ans, total, res)
                
                total -= number
                ans.pop()